app/services/encrypted_redis_struct_storage.rb

```diff
@@ -0,0 +1,89 @@
+# Include this mixin in a plain Struct to add ability to store it in redis
+# Call +configure_encrypted_redis_struct+ with configuration values
+#
+# @example
+#   MyStruct = Struct.new(:id, :a, :b) do
+#     include EncryptedRedisStructStorage
+#     configure_encrypted_redis_struct key_prefix: 'myprefix'
+#   end
+#
+#   MyStruct.store(id: '123', a: 'a', 'b')
+#   s = MyStruct.load('123')
+module EncryptedRedisStructStorage
```

Comment:
I really like the structure and consolidation, but I feel like it may fit better as a parent class than as a mixin for structs?

`init_fields` and `configure_encrypted_redis_struct` are specifically what I think is pushing me in that direction

e.g. `redis_key_prefix` would be defined on child classes, `expires_in` has a default definition on the parent that can be overridden, and `init_fields` becomes a standard `initialize`. What do you think?

Reply:
There's not an easy way to subclass structs and add new fields, right?

Reply:
Another option would be a completely separate service object, no subclassing/mixing in:

```ruby
RedisStructStorage.store(struct_instance)
RedisStructStorage.load(id, type: MyStructType)
```

which isn't *that* much more verbose

Reply:
yeah, it couldn't use structs unfortunately

I had something along the lines of this in my head:

```ruby
class ProofingDocumentCaptureSessionResult < EncryptedRedisStorage
  attr_accessor :id, :pii, :result

  def redis_key_prefix
    'dcs-proofing:result'
  end

  def payload
    {
     'pii' => pii,
     'result' => result,
    }.to_json
  end
end

class EncryptedRedisStorage
  def key
    [redis_key_prefix, id].join(':')
  end
  
  def store
    REDIS_POOL.with do |client|
      client.write(
        key,
        Encryption::Encryptors::SessionEncryptor.new.encrypt(payload),
        expires_in: expires_in,
      )
    end
  end

  def load
    ciphertext = REDIS_POOL.with { |client| client.read(key) }
    return nil if ciphertext.blank?
    decrypt_and_deserialize(id, ciphertext)
  end

  def expires_in
    60
  end
  # ...
end
```

Reply:
I think `Struct` gives us a lot
- an easy way to enumerate properties (so we have to add new keys in multiple places, ex `attr_accessor` and `payload)
- the hash setting syntax `struct[key] = value`, so we have to do `send("#{attr}=", value)` which is okay, but not as clear

I also think there's an argument to me made that subclassing a base class violates the Liskov substitution part of SOLID, meaning we can't use a child class as a direct replacement for the parent, because in our case, because our goal here is to create a series of value objects that are not substitutes for each other, and the parent class is essentially abstract

(Not that we have a policy of sticking to SOLID principles in our codebase, but I like the general direction and try to keep it in mind when adding code)

Reply:
That's a good point. Given that, I think the service class is a solid direction to go in 🙂

Reply:
Ok! Gave it a shot in e25a09d6664576d80c8837bcb994e02728a7ed4c, PTAL!